OpenSpace/OpenSpace
# Sonification/Telemetry

# Telemetry and OSC modules
This PR adds two modules, one to send OSC messages and one to monitor the state of the engine/scene and send that out using the OSC module. This has been used to send scene information to another software, SuperCollider, that has then used the information to create a sonification of the planets in different aspects. However, the thought is that any software capable of receiving OSC messages can use this information for anything.

The goal with this module is that it is general enough that other sonifications can be created without us needing to implement anything extra on the OpenSpace side. It will also be a way to distribute the sonifications we have already created so that anyone can use them. 

Note that the Sonification module has been renamed to Telemetry to make it clearer that we do not produce the sonifications in OpenSpace, but rather, we send out data via OSC to a sonification software that, in turn, creates the sonification. This means that to run this and test it, you need to download another software [SuperCollider](https://supercollider.github.io/). 

Since I am starting to write the documentation for this module, I would appreciate any comments or suggestions to make anything clearer. Especially the _TelemetryModule.h_ file and the util files for the module contain quite a bit of documentation text that could be good to get other eyes on. 

## Steps to test it:
1. You need to build OpenSpace in CMake with the OSC and Telemetry module enabled 
2. Download and install [SuperCollider](https://supercollider.github.io/)
3. Run the SuperCollider test file _OpenSpace-OSC-example.scd_ located in _data\assets\examples\sonification_.
    - To run the file in SuperCollider, click any line inside any OSC definitions (for example, line 31) and press CTRL + ENTER.
    - The SuperCollider console should respond with: `-> OSCdef(Neptune, /Neptune, nil, nil, nil).`
4. Run OpenSpace with the sonification profile.
5. To test the general nodes sonification, you need to drag and drop the asset file _data\assets\examples\sonification\nodes.asset_
6. Turn on the Telemetry module under _Settings/Modules_ in the GUI
7. Turn on all the telemetries/sonifications that are separate checkboxes in the telemetry module settings
8. Fly around in OpenSpace, and you should see messages being printed in the SuperCollider console window
9. Now, you can freely try the different settings for the different telemetries/sonifications, read the tooltips, and stress-test it in general.
10. **Before you shut down OpenSpace, you will need to turn off the telemetry module in the settings menu.** If OpenSpace is shut down while the telemetry module is still enabled, it will crash. This is an issue with the Scene not being safe with multiple threads, especially around startup and shutdown. 

## Things that are still in progress:
* Adjusting the OpenSpace planet SuperCollider sonification code to the new OSC structure and adding it into the repository, the plan is to put it in _data\assets\modules\telemetry\sonification_ 
* Write extensive documentation about the new module, especially about the math and the different angle modes
* Add a scaled-down sonification example for the general nodes sonification
